cpp/src_prims/sparse/coo.h

```diff
@@ -630,13 +642,13 @@ void coo_remove_scalar(COO<T> *const in, COO<T> *out, T scalar,
  * @param stream: cuda stream to use
  */
 template <int TPB_X, typename T>
-void coo_remove_zeros(COO<T> *const in, COO<T> *out, cudaStream_t stream) {
+void coo_remove_zeros(const COO<T> *in, COO<T> *out, cudaStream_t stream) {
   coo_remove_scalar<TPB_X, T>(in, out, T(0.0), stream);
 }
 
 template <int TPB_X, typename T>
-__global__ void from_knn_graph_kernel(long *const knn_indices,
-                                      T *const knn_dists, int m, int k,
+__global__ void from_knn_graph_kernel(const long *knn_indices,
```

Comment:
Another custom kernel.

Reply:
This turns a knn graph into a sparse data structure. What do you mean by custom kernel? It’s a sparse primitive.